x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java

```diff
@@ -0,0 +1,281 @@
+            .map(t -> MlTasks.jobId(t.getId()))
+            .collect(Collectors.toSet());
+        Set<String> waitingAnalysisJobs = dataframeAnalyticsTasks.stream()
+            .filter(t -> AWAITING_LAZY_ASSIGNMENT.equals(t.getAssignment()))
+            .map(t -> MlTasks.dataFrameAnalyticsId(t.getId()))
+            .collect(Collectors.toSet());
+        anomalyJobsTimeInQueue.keySet().retainAll(waitingAnomalyJobs);
+        analyticsJobsTimeInQueue.keySet().retainAll(waitingAnalysisJobs);
+
+        if (waitingAnomalyJobs.isEmpty() == false || waitingAnalysisJobs.isEmpty() == false || nodes.size() < decider.getMinNumNodes()) {
+            if (nodes.size() < decider.getMinNumNodes() || nodes.isEmpty()) {
+                return new AutoscalingDecision(name(),
+                    AutoscalingDecisionType.SCALE_UP,
+                    "number of machine learning nodes ["
+                        + nodes.size()
+                        + "] is below the configured minimum number of ["
+                        + decider.getMinNumNodes()
+                        + "] or is zero");
+            }
+            Set<String> timedUpJobs = new HashSet<>();
+            for (String jobId : waitingAnomalyJobs) {
+                long time = anomalyJobsTimeInQueue.compute(jobId, (k, v) -> v == null ? 0L : v + timeSinceLastCheckMs);
+                if (time >= decider.getAnomalyJobTimeInQueue().getMillis()) {
+                    timedUpJobs.add(jobId);
+                }
+            }
+            for (String jobId : waitingAnalysisJobs) {
+                long time = analyticsJobsTimeInQueue.compute(jobId, (k, v) -> v == null ? 0L : v + timeSinceLastCheckMs);
+                if (time >= decider.getAnalysisJobTimeInQueue().getMillis()) {
+                    timedUpJobs.add(jobId);
+                }
+            }
+            if (timedUpJobs.isEmpty() == false) {
+                return new AutoscalingDecision(name(),
+                    AutoscalingDecisionType.SCALE_UP,
+                    "jobs " + timedUpJobs + " have been waiting for assignment too long");
+            }
+        }
+        return new AutoscalingDecision(name(), AutoscalingDecisionType.NO_SCALE, "no jobs have waited long enough for assignment");
+    }
+
+    AutoscalingDecision checkForScaleDown(MlAutoscalingDeciderConfiguration decider,
+                                          List<DiscoveryNode> nodes,
+                                          ClusterState clusterState) {
+        if (nodes.size() == decider.getMinNumNodes()) {
+            return new AutoscalingDecision(name(),
+                AutoscalingDecisionType.NO_SCALE,
+                "| already at configured minimum node count ["
+                    + decider.getMinNumNodes()
+                    +"]");
+        }
+        // kick of a refresh if our data around memory usage is stale.
+        if (nodeLoadDetector.getMlMemoryTracker().isRecentlyRefreshed() == false) {
```

Comment:
The definition of `isRecentlyRefreshed()` is currently "in the last 90 seconds".  We should probably change that so it's greater than the expected interval at which the autoscaling decider is called.  Otherwise we'll get situations where the memory tracker is stale on every single call to autoscaling.  I expect this won't happen as we scale up, because we refresh the memory tracker when attempting to open jobs.  But for scaling down when jobs are being closed and no new jobs opened we could easily get into the state where we repeatedly say "no scale" because the memory tracker is stale.  Maybe there is a way to tie the two intervals together in code instead of hardcoding `MlMemoryTracker.RECENT_UPDATE_THRESHOLD` to some value in the Cloud code.

Reply:
I see two possibilities:

- Have this aging rate configurable via a setting. This way consumers of this API can set it to some value larger than the polling interval
- Have this refresh rate tied to some value included in the polling itself. Initially, if it is stale given the interval of 90 sec, we say its stale. Then we keep track of the last API call time and if we have refreshed since that time (barring any other changes), we say we have fresh data.

I think the second option is doable and the nicest with very little configuration for the user. We will just have to think about what it means for job values to be stale and scaling down.